Reduce particle creations in load_tracers

gassmoeller opened the pull request:
This is the first of a series of pull requests optimizing some aspects of particle transfer. This one specifically:
- Adds a move constructor to particles that are generated from other particles that are not longer needed. @bangerth: I am not sure if what I implemented in the constructor is already optimal, but at least it is four times faster than the copy constructor (no more allocation of dynamic memory). Is there something I am missing?
- Reduces the number of particle constructor calls from `load_tracers` and if possible use the move-constructor
- Provides insertion hints inside `load_tracers` into the particles multimap (because all tracers of the same cell are inserted into the same region of the multimap), which improves the scaling behaviour of this function (but does not matter for models with few cells per process.

bangerth commented on include/aspect/particle/particle.h:
```diff
@@ -141,11 +141,29 @@ namespace aspect
          */
-        virtual
-        ~Particle ();
+        Particle (const Particle<dim> &particle);
```
Am I correct that you need to declare the regular copy constructor here because that's required if you *also* have a move constructor, but that you don't need it if not in C++11 because then you can just let the compiler generate the default copy constructor?

tjhei commented:
> Is there something I am missing?

Looks good to me. I am not sure if move assignments should check for self-assignment like copy assignment operators should, but maybe it wouldn't hurt to put in assert in?

bangerth commented on source/particle/particle.cc:
```diff
@@ -76,11 +76,47 @@ namespace aspect
+      :
+      location (particle.location),
+      reference_location(particle.reference_location),
+      id (particle.id),
```
If this is a non-trivial data object, you probably want to also use `std::move(particle.id)` here. This may require the addition of a move constructor to `CellId`.

bangerth commented on source/particle/particle.cc:
```diff
@@ -76,11 +76,47 @@ namespace aspect
+      location = particle.location;
+      reference_location = particle.reference_location;
+      id = particle.id;
+      properties.swap(particle.properties);
```
that's one way to do it. I think you can also do
```
  properties = std::move(particle.properties);
```
which may be more obvious.

bangerth commented on source/particle/particle.cc:
```diff
@@ -76,11 +76,47 @@ namespace aspect
     {
+      location = particle.location;
+      reference_location = particle.reference_location;
+      id = particle.id;
```
same as for the move constructor -- try moving it.

bangerth commented on source/particle/world.cc:
```diff
@@ -516,30 +516,77 @@ namespace aspect
+                                               std::make_pair(std::make_pair(cell->level(),cell->index()),
+                                                              Particle<dim>(pdata,property_manager->get_particle_size())));
+#endif
+              ++position_hint;
```
so you just move the hint forward by one here, in anticipation of the next particle being inserted just after this one. that's correct of the next particle resides on the same cell. is there anything we could do better if we find out that the next particle is not on the same cell? (Just asking out of curiosity.)

bangerth commented:
> Looks good to me. I am not sure if move assignments should check for self-assignment like copy assignment operators should, but maybe it wouldn't hurt to put in assert in?

See, this is interesting, because I asked myself that recently as well. I think that this is generally not possible. It certainly isn't possible for the move constructor, because the object that is being constructed can not already be the initializer. [1] 

For move copy operators, it's only a move operation if the rhs is an rvalue-reference, which is typically a temporary object that results from another operation. So when you write
```
   object = object;
```
then the rhs is an lvalue-reference. You'd have to explicitly write something like
```
  object = std::move(object);
```
which I think is not what one does. So I don't think that the assertion is actually necessary.


[1] It would be possible for the move constructor if you wrote this ill-formed and ill-conceived construction:
```
  class Base {
     Base (Base &&);   // move constructor
  };

  class Derived : public Base {
     Derived (const Derived &d)
       :  Base (std::move(*this))        // calls move constructor on itself
    {}
```

gassmoeller commented:
I have added a self assignment check for the assignment operators (found many cases where people argue to use them), but instead of throwing an exception, which may be slow, it simply does nothing if the two objects are the same. I think this is reasonable since the following code should be valid:
```
Particle<dim> a;
Particle<dim> &b = &a;
b = std::move(a);
```
It simply does not do anything useful, but it does not hurt either.
About the other comments:
- Yes I need to declare the copy constructor and the assignment operators when I declare a move constructor (although this seems to depend a bit on the compiler, gcc-4.8.4 on my laptop only requires the copy constructor, gcc-5.1 on our cluster also requires the assignment operators, a bit weird). In the non CXX11 case copy constructor and assignment is created by the compiler.
- particle.id is a trivial object (just an unsigned int or and unsigned long long), it is not a CellId object, so moving it is probably not faster than copying. What is your thought about `Tensor` or `Point` objects? They own statically allocated arrays, would they benefit from a move constructor?
- Yes I advance the hint by one since the next element with the same key in a multimap is inserted at the end of the existing elements with the same key (and the hint should point at the first element which goes after the element to insert). I have no idea how to optimize this when the next particle is not in the same cell. Luckily this only happens when cells are refined (persisting or coarsening throws all particles into one cell). Also I guess that children of one cell are relatively 'close' to each other in the multimap, because it is sorted by (cell_level,cell_index) so the hint is nevertheless a good guess. Do children of a cell get consecutive cell indices? All in all it is rare that we insert particles randomly into unrelated cells, it mostly happens after the particles moved (in the 'sort' stage).

I think I addressed all comments if the tester is happy.

bangerth commented:
About moving `particle.id` as well as points and tensors: for all statically sized objects, copying and moving is the same -- there is no representation that you can steal. So copying is ok.

Regarding the hint -- thanks for thinking this through. The children of a cell are typically consecutive (except in the case of anisotropic refinement, in which case we only guarantee that pairs of children are consecutive, but not necessarily the members of the 2 or 3 pairs of children).

bangerth commented:
Good work -- thanks!
